opt: strongly type Case inputs

RaduBerinde opened the pull request:
Currently, values for various branches in a Case can be of UnknownType. When we
infer the type of Case, we pick the first branch that has a definite type. This
is fragile: normalization rules can remove branches, and we can lose the only
branch with a definite type, changing the type of the Case itself.

This change makes sure all inputs to Case are always typed consistently, even if
it means adding casts. In practice, this mostly means that we use properly typed
Null operators.

Fixes #47299.

Release note (bug fix): fixed internal errors in some queries involving CASE
expressions with NULL branches.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50626)
<!-- Reviewable:end -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 14 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50626#-:-MAmK5s674lIXZ8zZ-mf:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball and @mgartner)



<!-- Sent from Reviewable.io -->

RaduBerinde committed:
opt: strongly type Case inputs

Currently, values for various branches in a Case can be of UnknownType. When we
infer the type of Case, we pick the first branch that has a definite type. This
is fragile: normalization rules can remove branches, and we can lose the only
branch with a definite type, changing the type of the Case itself.

This change makes sure all inputs to Case are always typed consistently, even if
it means adding casts. In practice, this mostly means that we use properly typed
Null operators.

Fixes #47299.

Release note (bug fix): fixed internal errors in some queries involving CASE
expressions with NULL branches.

RaduBerinde commented:
bors r+
